src/connectivity.jl

```diff
@@ -1,26 +1,32 @@
 # Parts of this code were taken / derived from Graphs.jl. See LICENSE for
 # licensing details.
 """
-    connected_components!(label, g)
+    connected_components!(label, g, [search_queue])
 
 Fill `label` with the `id` of the connected component in the undirected graph
 `g` to which it belongs. Return a vector representing the component assigned
 to each vertex. The component value is the smallest vertex ID in the component.
 
+A `search_queue`, an empty `Vector{eltype(edgetype(g))}`, can be provided to reduce
+allocations if `connected_components!` is intended to be called multiple times sequentially.
+If not provided, it is automatically instantiated.
+
 ### Performance
 This algorithm is linear in the number of edges of the graph.
 """
-function connected_components!(label::AbstractVector, g::AbstractGraph{T}) where {T}
+function connected_components!(
+    label::AbstractVector{T}, g::AbstractGraph{T}, search_queue::Vector{T}=Vector{T}()
+) where {T}
+    isempty(search_queue) || error("provided `search_queue` is not empty")
```

Comment:
If we pass in `search_queue`  to increase performance then it gets modified anyways. So instead of requiring that it is empty we can either call `empty!` on it. Calling `empty!` does not always (or maybe even never) deallocate the memory buffer that is associated with a `Vector` - so `push!` would not do unnecessary allocations when that vector is filled again.

Reply:
Agree, yeah, that's neater. Done.